created article What Defines a Middleware Architecture? (#262)
laminas/getlaminas.org: data/blog/2025/2025-03-18-mezzio101-what-defines-a-middleware-architecture.md

---
id: 2025-03-18-mezzio101-what-defines-a-middleware-architecture
author: bidi
title: 'Mezzio101: What Defines a Middleware Architecture?'
draft: false
public: true
created: '2025-03-18T11:00:00-01:00'
updated: '2025-03-18T11:00:00-01:00'
tags:
  - middleware
  - software architecture
  - framework
openGraphImage: '2025-03-18-mezzio101-what-defines-a-middleware-architecture.png'
openGraphDescription: 'Mezzio101: What Defines a Middleware Architecture?'
---

Nowadays, applications are getting more and more complex.

> The classic way of executing sequential code, from **request to response**, still works for simple apps, but using a modern architecture has many benefits.

Reviewer: This may need to be expanded. When you say this, do you mean a single PHP script that accepts a request and outputs the information, like we used to do in the PHP 4 days, or do you mean an MVC architecture - which is hardly sequential?

> Enter the **middleware**, an intermediary code that processes the request in various ways before it even reaches your custom code.

Reviewer: ```suggestion
Enter **middleware**: intermediary code that processes the request and/or response in various ways before it even reaches your custom code.
```

I think it's important to highlight that it can also modify the response.

<!--- EXTENDED -->

> In this article we will list the advantages of using middlewares in your applications.

Reviewer: ```suggestion
In this article we will list the advantages of using middleware in your applications.
```

The plural of "middleware" is "middleware"; please review and correct throughout.

From maintainability, to future-proof design, the benefits can't be ignored.

### What does middleware do?

The versatility of middleware is virtually endless.

- Do you need it to pre-process your request by validating the user session or the form input?
- Do you need to cache or translate the request?
- Do you need to contact a third party site to enhance the request?

Your custom middleware can handle **any specific logic** your application may require.
Some useful things to implement are error handling, CORS, routing, translating.

Ultimately, you can split your application into multiple middlewares that process the request before it reaches your main code.
An added benefit is that the middlewares can be **reused more easily** in other applications.

> Even better, the middlewares **don't need to be written in the same programming language** or coding architecture.

Reviewer: This one is ... not quite right.

There are two types of middleware architectures:

- One used within a framework to act as a transformative pipeline to transform a request to a response (that's where Mezzio sits)
- One that acts as a filter/translation/router for data, ultimately determining if data needs to be handled, and, if so, where to funnel it (this is the realm of Kafka)

I'd likely remove this sentence, or provide these clarifications.

Author: @weierophinney , i hope it's ok to add your comment in there. please let me know if you want me to edit anything

> The alternative to the middleware architecture is the MVC architecture which has grown considerably since its inception, is feature rich, modular and reusable.

Reviewer: ```suggestion
One alternative to the middleware architecture is the MVC architecture which has grown considerably since its inception, is feature rich, modular and reusable.
```

Instead of using flexible layers like a middleware, MVC revolves around controllers and actions.

> Unfortunately, MVC's complexity has brought about its downfall, though it is still being used.

Reviewer: "downfall" might be an overstatement; it's still heavily used (Spring, Laravel, Rails, etc.). I'd remove this sentence.

> The important bit and the central difference to middleware is that in a MVC, a request targets a single controller.

Reviewer: ```suggestion
The primary difference to middleware is that in an MVC, a request targets a single controller.
```

Also, expand on this - why is this a key differentiator? What does it enable? What does it make harder?

> Nowadays, the aim is more toward lightweight, modular and standardized frameworks for producing microframeworks and API based solutions which are more suited to a middleware approach.

Reviewer: ```suggestion
Middleware enables lightweight, modular, and standardized frameworks, which are particularly suited for functions like APIs.
```

### How do you set up a middleware pipeline?

The **middleware pipeline** defines the sequence in which the middlewares are executed.
Setting up the pipeline needs to be done with care to avoid unwanted behavior.
In some cases, the execution of one middleware can affect another middleware because of the way the request is passed along.
The main things to consider when setting up the pipeline are:

- Security
- Performance
- Functionality

For example, it makes sense to add an error handler early in the pipeline, even as the first entry.
If your application requires the user to be logged in, it doesn't make sense to process the request before the authentication middleware gives the go-ahead.
Also, it doesn't make sense to have an authorization middleware placed before an authentication middleware in the pipeline.

> Based on the future needs of your application, the **pipeline can be easily updated** to include additional middleware.

Reviewer: Explain this in more detail: why is it easier? Maybe provide an example, and potentially show how the same functionality in an MVC either requires deep ties into the MVC framework, or functionality that cannot be re-used.

Author: i expanded this section with an example and more on the MVC alternative

This aspect goes a long way toward making the flow of your application easier to understand and debug.
After all, debugging and maintaining smaller sections of code is easier than working with a monolith.

### When and how is middleware executed?

Like the name suggests, **middlewares are executed between the request and the server response** handled by your main codebase.
The classic way of handling a request meant running through a sequence of code that is less adaptive.

> That is, unless you add a bunch of `if-else` statements for all the cases where the execution needs to exit prematurely.

Reviewer: This is... not really how most MVC frameworks work anymore. They tend to have lifecycle events that allow you to tie into different points of request handling, allowing early exit. It's not really if/else.

Author: my inital goal for this article was not really to compare middleware with MVC, but i guess the two are connected now. 

this particular paragraph compares middleware with the classic PHP4 architecture. in reality, it was intended to mock the old way of doing things, before MVC and middleware. so maybe the best solution is to remove the sentence with 'if-else' altogether?

Reviewer: Yes, please, @bidi47

---
id: 2025-03-18-mezzio101-what-defines-a-middleware-architecture
author: bidi
title: 'Mezzio101: What Defines a Middleware Architecture?'
draft: false
public: true
created: '2025-03-18T11:00:00-01:00'

> updated: '2025-03-18T11:00:00-01:00'

Reviewer: Make sure you update these once you're ready to publish!

### How do you set up a middleware pipeline?

The **middleware pipeline** defines the sequence in which the middleware are executed.
Setting up the pipeline needs to be done with care to avoid unwanted behavior.
In some cases, the execution of one middleware can affect another middleware because of the way the request is passed along.
The main things to consider when setting up the pipeline are:

- Security
- Performance
- Functionality

For example, it makes sense to add an error handler early in the pipeline, even as the first entry.
If your application requires the user to be logged in, it doesn't make sense to process the request before the authentication middleware gives the go-ahead.
Also, it doesn't make sense to have an authorization middleware placed before an authentication middleware in the pipeline.

Based on the future needs of your application, the **pipeline can be easily updated** to include additional middleware.
Consider this practical example:

- Say you begin with a simple presentation site that doesn't require a lot of bells and whistles.

> Your main focus is on templating and displaying mostly-static pages.

Reviewer: You may need to indent this line so that it stays with the bullet point.

Author: thank you for the feedback, @weierophinney 

the text is properly aligned under the bullet points in my editor, but i'll change this if the preview doesn't look good

Reviewer: I was under the impression that subsequent lines for bullet points needed to be indented, but when you look [at the preview on GitHub](https://github.com/bidi47/getlaminas.org/blob/mezzio-101-middleware/data/blog/2025/2025-04-08-mezzio101-what-defines-a-middleware-architecture.md), it renders fine. That said, I think if you wanted multiple paragraphs for the bullet, or to add something like a code block, you'd need the indentation — which is likely why I'm in the habit of indenting, as it means that all bullets are then consistently formatted.

- Later on, you expand your application with a blog that allows user comments.

> You suddenly require authentication, so you add it as a new layer in the middleware pipeline.

Reviewer: Same here, and throughout this list.

> - Later still, you define different level of access to your users via authorization.

Reviewer: ```suggestion
- Later still, you define different levels of access to your users via authorization.
```
